node/src/exchange/server.rs

```diff
@@ -1,52 +1,396 @@
+use std::fmt::{Debug, Display};
+use std::future::Future;
+use std::pin::Pin;
 use std::sync::Arc;
+use std::task::{Context, Poll};
 
-use celestia_proto::p2p::pb::{HeaderRequest, HeaderResponse};
+use celestia_proto::p2p::pb::{header_request, HeaderRequest, HeaderResponse};
+use celestia_types::Hash;
+use futures::stream::FuturesUnordered;
+use futures::{FutureExt, Stream};
 use libp2p::{
     request_response::{InboundFailure, RequestId, ResponseChannel},
     PeerId,
 };
-use tracing::instrument;
+use tendermint::hash::Algorithm;
+use tracing::{instrument, trace};
 
+use crate::exchange::utils::{ExtendedHeaderExt, HeaderRequestExt, HeaderResponseExt};
+use crate::exchange::{ReqRespBehaviour, ResponseType};
 use crate::store::Store;
 
-pub(super) struct ExchangeServerHandler<S>
+type StoreJobType<C> = dyn Future<Output = (C, ResponseType)> + Send;
+
+pub(super) struct ExchangeServerHandler<S, R = ReqRespBehaviour>
 where
-    S: Store + 'static,
+    S: Store,
+    R: ResponseSender,
 {
-    _store: Arc<S>,
+    store: Arc<S>,
+    store_jobs: FuturesUnordered<Pin<Box<StoreJobType<R::Channel>>>>,
+}
+
+pub(super) trait ResponseSender {
+    type Channel: Send + 'static;
+
+    fn send_response(&mut self, channel: Self::Channel, response: ResponseType);
+}
+
+impl ResponseSender for ReqRespBehaviour {
+    type Channel = ResponseChannel<ResponseType>;
+
+    fn send_response(&mut self, channel: Self::Channel, response: ResponseType) {
+        // response was prepared specifically for the request, we can drop it
+        // in case of error we'll get Event::InboundFailure
+        let _ = self.send_response(channel, response);
+    }
 }
 
-impl<S> ExchangeServerHandler<S>
+impl<S, R> ExchangeServerHandler<S, R>
 where
     S: Store + 'static,
+    R: ResponseSender,
 {
     pub(super) fn new(store: Arc<S>) -> Self {
-        ExchangeServerHandler { _store: store }
+        ExchangeServerHandler {
+            store,
+            store_jobs: FuturesUnordered::new(),
+        }
     }
 
-    #[instrument(level = "trace", skip(self, _respond_to))]
-    pub(super) fn on_request_received(
+    #[instrument(level = "trace", skip(self, response_channel))]
+    pub(super) fn on_request_received<ID>(
         &mut self,
         peer: PeerId,
-        request_id: RequestId,
+        request_id: ID,
         request: HeaderRequest,
-        _respond_to: ResponseChannel<Vec<HeaderResponse>>,
-    ) {
-        // TODO
+        response_channel: R::Channel,
+    ) where
+        ID: Display + Debug,
+    {
+        let Some((amount, data)) = self.parse_request(request) else {
+            self.store_jobs
+                .push(handle_invalid_request(response_channel).boxed());
+            return;
+        };
+
+        let store_job = match data {
+            header_request::Data::Origin(0) => {
+                handle_request_current_head(self.store.clone(), response_channel).boxed()
+            }
```

Comment:
I prefer standalone functions only when they are an independent utility or when there is partial borrowing issue. 

You could define the handle method like this:

```rust
    fn handle_request_current_head(&mut self, channel: R::Channel) {
        let store = self.store.clone();

        let fut = async move {
            let response = store
                .get_head()
                .await
                .map(|head| head.to_header_response())
                .unwrap_or_else(|_| HeaderResponse::not_found());

            (channel, vec![response])
        };

        self.store_jobs.push(fut.boxed());
    }
```

and here use it like this:

```rust
            header_request::Data::Origin(0) => {
                self.handle_request_current_head(response_channel);
            }
```

which IMO will result to cleaner code

Reply:
I've started with named utility functions to better convey what's being done to the store, but this indeed results in cleaner code and less generics, thanks!